va/va_prot.h

```diff
@@ -0,0 +1,379 @@
+ * wired = attrib_prot[0].value;
+ * VAConfigAttribDisplayLinkHdcpWireless = attrib_prot[0].value;
+ * if ((wired.bits.session_type & VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0) == 0) {
+ *     // not find desired mode
+ *     assert(0);
+ * }
+ * \endcode
+ *
+ * \section api_prot_setup Set up a protected session
+ *
+ * A protected session is a session between application and GPU hardware.
+ * This protected session can be attached to the display. Also with protected session,
+ * application can do specific TEE execution functionality with TEE inside GPU, but it depends
+ * on IHV's implementation.
+ * Before creating a protected session, it needs to create a config first via
+ * vaCreateConfig(). Then using this config id to create a protected session via
+ * vaCreateProtectedSession().
+ *
+ * The general control flow is demonstrated by the following pseudo-code:
+ * \code
+ * // Create config
+ * VAConfigID config_id;
+ *
+ * wired.bits.session_type = VA_PROTECTED_DISPLAY_LINK_SESSION_TYPE0;
+ * wired.bits.protocol_version = VA_PROTECTED_DISPLAY_LINK_HDCP_VERSION_22;
+ * attrib_prot[0].value = wired.value;
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected, VAEntrypointProtectedDisplayLink, attrib_prot, 1, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected session via vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID protected_session;
+ *
+ * va_status = vaCreateProtectedSession(
+ *     va_dpy,
+ *     config_id,
+ *     &protected_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_prot_attach Attach/Detach protected session to the context which
+ * want to enable/disable protection
+ *
+ * Protected session is attached to decode/encode/vp context to enable protection.
+ *
+ * \code
+ * vaAttachProtectedSession(va_dpy, decode_ctx, protected_session);
+ * foreach (iteration) {
+ *     vaBeginPicture(va_dpy, decode_ctx, surface);
+ *     ...
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id1, 1);
+ *     vaRenderPicture(va_dpy, decode_ctx, &buf_id2, 1);
+ *     ...
+ *     vaEndPicture(va_dpy, decode_ctx);
+ * }
+ * vaDetachProtectedSession(va_dpy, decode_ctx);
+ * \endcode
+ *
+ * \section api_prot_update Updating protected session status related to
+ * hardware settings
+ *
+ * A function for updating hardware status of the protected session. Sometimes
+ * the status is initialized at protected session creation time. When the
+ * information is gathered, we can update the protected session status based
+ * on the information we got with vaProtectedSessionHwUpdate().
+ * For example, if the protected session is for protected content, we will get
+ * policy blobs from TEE and then pass these infomation with
+ * vaProtectedSessionHwUpdate to change the session status to enable specific
```

Comment:
As I noted few times in this review I still don't fully understand call sequence. So, how the process ultimately look like? Is that like this:
```
vaCreateProtectedSession();
vaProtectedSessionExecute(); // get supported policies
vaProtectedSessionHwUpdate(); // apply selected policy
vaAttachProtectedSession();
vaRenderPicture -> vaRenderPicture -> vaEndPicture
vaDetachProtectedSession()
```

Is that like that or looks differently? Also, you actually have output buffer for the HwUpdate... which kind of feedback you can get thru it and how you will use it?

Reply:
Thanks for your feedback. I restructure and refine the comments and updated. Please check if it is better or not.